File: common/circles.ts
import dayjs, { Dayjs } from 'dayjs';
import localizedFormat from 'dayjs/plugin/localizedFormat';
dayjs.extend(localizedFormat);

import canvas, { loadImage } from 'canvas';
import { InteractionsType, SimpleProfileType } from './types';
import { DO_NOT_INCLUDE_THESE } from './defaults';

function hex_is_light(color: string) {
    const hex = color.replace('#', '');
    const c_r = parseInt(hex.substring(0, 0 + 2), 16);
    const c_g = parseInt(hex.substring(2, 2 + 2), 16);
    const c_b = parseInt(hex.substring(4, 4 + 2), 16);
    const brightness = (c_r * 299 + c_g * 587 + c_b * 114) / 1000;
    return brightness > 155;
}
const toRad = (x: number) => x * (Math.PI / 180);

export const createCirclesImage = async (
    profile: SimpleProfileType,
    data: { [key: string]: InteractionsType[] },
    date: { type: string; start?: Dayjs; end?: Dayjs },
    locale: string | undefined
) => {
    const imageSize = 900;
    const cv = canvas.createCanvas(imageSize, imageSize);
    const context = cv.getContext('2d');

    let peopleList: SimpleProfileType[] = [];

    if (!locale) locale = 'en';

    const options = {
        orbits: 2,
        include_sent: true,
        include_rcvd: false,
        remove_bots: true,
        add_watermark: true,
        add_date: true,
        bg_color: '#1D428A',
        add_border: true,
        border_color: '#FFC72C'
    };

    // decides which interactions to use based on the options
    let interactionsList: InteractionsType[] | undefined;
    if (options.include_sent && options.include_rcvd) {
        interactionsList = data.both;
    } else if (options.include_sent) {
        interactionsList = data.sent;
    } else if (options.include_rcvd) {
        interactionsList = data.rcvd;
    }

    // no image if no data
    if (!interactionsList || interactionsList.length === 0) return;

    // filter bots
    if (options.remove_bots) {
        interactionsList = interactionsList.filter((x) => !DO_NOT_INCLUDE_THESE.includes(x._id));
    }

    // - radial distances for each number of orbits
    // - i chose this manually
    const distances: { [key: number]: number[] } = {
        1: [0, 1/3 + 2/100, 0, 0],
        2: [0, 1/4 + 1/200, 3/8 + 3/100, 0],
        3: [0, 1/5 + 1/200, 5/16 + 2/100, 7/16 + 1/300]
    };

    // radiuses for every orbit for each number of orbits
    const radiuses: { [key: number]: number[] } = {
        1: [20/100, 10/100, 0, 0],
        2: [14/100, 7/100, 6/100, 0],
        3: [12/100, 6/100, 5/100, 4/100]
    };

    // - main input for the image generation later
    // - each block is an orbit, 0-th orbit is the
    //   main profile image
    let config = [
        { distance: 0, count: 1, radius: radiuses[options.orbits][0], users: [profile] },
        {
            distance: distances[options.orbits][1],
            count: 10,
            radius: radiuses[options.orbits][1],
            users: interactionsList.slice(0, 10).map((x) => x.profile)
        },
        {
            distance: distances[options.orbits][2],
            count: 20,
            radius: radiuses[options.orbits][2],
            users: interactionsList.slice(10, 30).map((x) => x.profile)
        },
        {
            distance: distances[options.orbits][3],
            count: 30,
            radius: radiuses[options.orbits][3],
            users: interactionsList.slice(30, 60).map((x) => x.profile)
        }
    ];
    config = config.slice(0, options.orbits + 1);

    const textColor = hex_is_light(options.bg_color) ? '#000000' : '#CCCCCC';

    context.fillStyle = options.bg_color;
    context.fillRect(0, 0, imageSize, imageSize);

    context.imageSmoothingEnabled = true;
    // context.imageSmoothingQuality = 'medium';

    const fontSize = imageSize/35;

    // date on top left corner
    if (options.add_date) {
        let textFull = '';
        if (date.type === 'weekly') {
            const textFrom = date.start?.toDate().toLocaleDateString(locale);
            const textTo = date.end?.toDate().toLocaleDateString(locale);
            textFull = `${textFrom} - ${textTo}`;
        } else if (date.type === 'all') {
            textFull = `${new Date().toLocaleDateString(locale)} (all time)`;
        } else if (date.type === 'month') {
            textFull = `${new Date().toLocaleDateString(locale)} (month)`;
        } else if (date.type === 'week') {
            const textFrom = dayjs().subtract(1, 'week').toDate().toLocaleDateString(locale);
            const textTo = new Date().toLocaleDateString(locale);
            textFull = `${textFrom} - ${textTo}`;
        } else if (date.type === 'day') {
            const textFrom = dayjs().subtract(24, 'hour').toDate().toLocaleDateString(locale);
            const textTo = new Date().toLocaleDateString(locale);
            textFull = `${textFrom} - ${textTo}`;
        }
        context.font = `${fontSize}px Arial`;
        context.fillStyle = textColor;
        context.textAlign = 'left';
        context.textBaseline = 'top';
        context.fillText(textFull, imageSize/50, imageSize/60);
    }

    // site watermark on top right corner
    if (options.add_watermark) {
        context.font = `${fontSize}px Arial`;
        context.fillStyle = textColor;
        context.textAlign = 'right';
        context.textBaseline = 'top';
        context.fillText('wolfgang.raios.xyz', imageSize - imageSize/50, imageSize/60);
    }

    // add rounded border
    if (options.add_border) {
        context.strokeStyle = options.border_color;
        context.lineWidth = imageSize/40;
        context.beginPath();
        context.roundRect(0, 0, imageSize, imageSize, imageSize/40);
        context.stroke();
    }
    const promises: Promise<any>[] = [];

    // this will create the image, load the avatar and return a promise
    const preload = (user: { [key: string]: string | undefined }, opt: { [key: string]: number }) =>
        new Promise(async (resolve, reject) => {
            if (!user.avatar) return reject();
            return loadImage(user.avatar)
                .then((img) => {
                    context.save();
                    // this draws a circle centered at the image position
                    context.beginPath();
                    context.arc(opt.centerX, opt.centerY, opt.radius, 0, 2 * Math.PI, false);
                    // then clips whatever is out
                    context.clip();
                    // this draws the img at some position with some radius
                    context.drawImage(
                        img,
                        opt.centerX - opt.radius,
                        opt.centerY - opt.radius,
                        opt.radius * 2,
                        opt.radius * 2
                    );
                    context.restore();
                    return resolve(img);
                })
                .catch(async () => {
                    return loadImage('./common/person-fill.svg')
                        .then((img) => {
                            context.save();
                            context.beginPath();
                            context.arc(
                                opt.centerX,
                                opt.centerY,
                                opt.radius,
                                0,
                                2 * Math.PI,
                                false
                            );
                            context.clip();
                            context.drawImage(
                                img,
                                opt.centerX - opt.radius,
                                opt.centerY - opt.radius,
                                opt.radius * 2,
                                opt.radius * 2
                            );
                            context.restore();
                            return resolve(img);
                        })
                        .catch(() => reject());
                });
        });

    // now we iterate the orbits to actually build the full image
    for (const [orbitIndex, orbit] of config.entries()) {
        const { count, radius, distance, users } = orbit;

        // number of slices in this orbit
        const angleSize = 360 / count;
        // iterate through all users in this orbit
        for (let i = 0; i < count; i++) {
            // if list ends here we stop
            if (!users[i]) break;

            // 30 degrees offset for every additional orbit
            const offset = orbitIndex * 30;
            // final angle for this user at this orbit
            const t = toRad(i * angleSize + offset);

            if (orbitIndex > 0) {
                peopleList.push(users[i]);
            }

            // push a new image loading thingy into the promises list
            // with the coordinates and radius for that circle
            promises.push(
                preload(
                    { avatar: users[i].avatar, did: users[i].did },
                    {
                        centerX: Math.cos(t) * imageSize * distance + imageSize / 2,
                        centerY: Math.sin(t) * imageSize * distance + imageSize / 2,
                        radius: imageSize * radius
                    }
                )
            );
        }
    }

    const stream = await Promise.allSettled(promises).then(() => cv.toBuffer());
    
    return {
        people: peopleList,
        image: stream,
    }
};

File: backend/interactions/circles.py
# @app.get("/circles")
# async def _circles(actor: str, source: Literal["sent", "rcvd", "both"] = "rcvd"):
#     handle, did = await _get_did(actor)
#     if did is None:
#         raise HTTPException(status_code=404, detail=f"user not found: {actor}")

#     logger.info(f"[circles] getting interactions: {handle}@{did}")
#     try:
#         start_date = datetime.datetime.now() - datetime.timedelta(days=7)
#         interactions_data = await _get_interactions(did, start_date)
#         image_interactions = _generate_image_interactions(interactions_data)
#     except Exception:
#         raise HTTPException(status_code=500, detail=f"error generating interactions {handle}@{did}")

#     logger.info(f"[circles] generating image: {handle}@{did}")

#     # Getting stuff for the image
#     main_profile = await _get_profile(did)
#     if main_profile is None:
#         raise

#     profiles = []
#     for person in image_interactions:
#         profile = await _get_profile(person["_id"])
#         profiles.append(
#             {
#                 "did": profile.did,
#                 "handle": profile.handle,
#                 "avatar": profile.avatar,
#             }
#         )

#     if len(profiles) <= 1:
#         return

#     main_profile_picture, all_profile_pictures = await _fetch_profile_pictures(main_profile, profiles)
#     output_image = _create_circles_image(main_profile_picture, all_profile_pictures, start_date)
#     # except:
#     # raise HTTPException(status_code=500, detail="error generating circles")
#     if output_image is None:
#         raise HTTPException(status_code=500, detail=f"error generating circles {handle}@{did}")

#     stream = BytesIO()
#     output_image.save(stream, format="png")
#     stream.seek(0)  # important here!
#     return responses.StreamingResponse(stream, media_type="image/png")

import asyncio
import datetime
import math
from collections import defaultdict
from io import BytesIO

import aiohttp
from PIL import Image, ImageDraw, ImageFont

from backend.types import Interaction


def _hex_is_light(color):
    """
    Determines if a given hex color is considered "light" based on its brightness.

    Args:
        color: The hex color string (e.g., "#RRGGBB").

    Returns:
        True if the color is light, False otherwise.
    """
    hex_color = color.replace("#", "")
    c_r = int(hex_color[0:2], 16)
    c_g = int(hex_color[2:4], 16)
    c_b = int(hex_color[4:6], 16)
    brightness = (c_r * 299 + c_g * 587 + c_b * 114) / 1000
    return brightness > 155


def _fib(n):
    golden_ratio = (1 + math.sqrt(5)) / 2
    val = (golden_ratio**n - (1 - golden_ratio) ** n) / math.sqrt(5)
    return int(round(val))


def _generate_image_interactions(
    interactions_data: dict[str, list[Interaction]],
    topk: int = 50,
) -> list[Interaction]:
    combined_interactions = defaultdict(lambda: {"l": 0, "r": 0, "p": 0, "c": 0, "t": 0})

    for data in interactions_data.values():
        for tmp in data:
            combined_interactions[tmp["_id"]]["l"] += tmp["l"]
            combined_interactions[tmp["_id"]]["r"] += tmp["r"]
            combined_interactions[tmp["_id"]]["p"] += tmp["p"]
            combined_interactions[tmp["_id"]]["c"] += tmp["c"]
            combined_interactions[tmp["_id"]]["t"] += tmp["t"]

    combined_interactions = [Interaction(_id=key, **value) for key, value in combined_interactions.items()]
    combined_interactions.sort(key=lambda x: x["t"], reverse=True)
    return combined_interactions[:topk]


async def fetch_image(session: aiohttp.ClientSession, profile: dict[str, str]) -> tuple:
    try:
        async with session.get(profile["avatar"]) as response:
            data = await response.read()
            image = Image.open(BytesIO(data))
            return image
    except Exception:
        width = 60
        height = 60
        line_color = "black"
        image = Image.new("RGB", (width, height), "white")
        draw = ImageDraw.Draw(image)
        draw.line((0, 0, width - 1, height - 1), fill=line_color, width=1)
        draw.line((width - 1, 0, 0, height - 1), fill=line_color, width=1)
        return image


async def _fetch_profile_pictures(main_profile, profiles):
    async with aiohttp.ClientSession() as session:
        main_profile_picture = await fetch_image(session, main_profile)
        tasks = [fetch_image(session, x) for x in profiles]
        all_profile_pictures = await asyncio.gather(*tasks)
    return main_profile_picture, all_profile_pictures


def _create_circles_image(main_profile_picture, all_profile_pictures, start_date: datetime.datetime):
    _CIRCLES_OPTIONS = {
        "orbits": 2,
        "include_sent": True,
        "include_rcvd": False,
        "remove_bots": True,
        "remove_blocked": True,
        "add_watermark": True,
        "add_date": True,
        "bg_color": "#1D428A",
        "add_border": True,
        "border_color": "#FFC72C",
    }

    # Radial distances for each number of orbits
    _CIRCLES_DISTANCES = {
        1: [0, 35 / 100, 0, 0],
        2: [0, 23 / 100, 38 / 100, 0],
        3: [0, 20 / 100, 32 / 100, 42 / 100],
    }

    # Radiuses for every orbit for each number of orbits
    _CIRCLES_RADIUSES = {
        1: [19 / 100, 9 / 100, 0, 0],
        2: [13 / 100, 7 / 100, 6 / 100, 0],
        3: [11 / 100, 6 / 100, 5 / 100, 4 / 100],
    }

    image_size = 1800
    border_radius = 12

    n_orbits = _CIRCLES_OPTIONS["orbits"]
    start_index = 0
    _CIRCLES_ORBITS_DEF = []
    for i in range(n_orbits):
        count = _fib(i + 6)
        end_index = start_index + count
        _CIRCLES_ORBITS_DEF.append(
            {
                "count": count,
                "distance": _CIRCLES_DISTANCES[n_orbits][i + 1],
                "radius": _CIRCLES_RADIUSES[n_orbits][i + 1],
                "pictures": all_profile_pictures[start_index:end_index],
            }
        )
        start_index = end_index

    # Creating image
    cv = Image.new("RGBA", (image_size, image_size))
    context = ImageDraw.Draw(cv)

    text_color = "#000000" if _hex_is_light(_CIRCLES_OPTIONS["bg_color"]) else "#CCCCCC"

    # Background color
    context.rounded_rectangle(
        [(0, 0), (image_size, image_size)], radius=border_radius, fill=_CIRCLES_OPTIONS["bg_color"]
    )

    # Rounded border
    if _CIRCLES_OPTIONS["add_border"]:
        context.rounded_rectangle(
            [(0, 0), (image_size, image_size)],
            radius=border_radius,
            outline=_CIRCLES_OPTIONS["border_color"],
            width=image_size // 60,
        )

    font_size = image_size // 30
    font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf", font_size)

    # Date on top left corner
    if _CIRCLES_OPTIONS["add_date"]:
        now = datetime.datetime.now()
        text_full = f"{start_date.strftime('%Y-%m-%d')} - {now.strftime('%Y-%m-%d')}"
        context.text((image_size / 35, image_size / 45), text_full, font=font, fill=text_color, anchor="la")

    # Site watermark on top right corner
    if _CIRCLES_OPTIONS["add_watermark"]:
        context.text(
            (image_size - image_size / 35, image_size / 45),
            "wolfgang.raios.xyz",
            font=font,
            fill=text_color,
            anchor="ra",
        )

    def add_picture_to_image(img: Image.Image, cfg: dict):
        width, height = img.size
        if width != height:
            size = min(width, height)
            left = (width - size) / 2
            top = (height - size) / 2
            right = (width + size) / 2
            bottom = (height + size) / 2
            img = img.crop((left, top, right, bottom))
        img = img.resize((int(cfg["r"] * 2), int(cfg["r"] * 2)), Image.Resampling.LANCZOS)

        # Create a circular mask
        mask = Image.new("L", (int(cfg["r"] * 2), int(cfg["r"] * 2)), 0)
        draw_mask = ImageDraw.Draw(mask)
        draw_mask.ellipse((0, 0, cfg["r"] * 2, cfg["r"] * 2), fill=255)

        # Apply the mask to the image

        img.putalpha(mask)

        # Paste the image onto the canvas
        cv.paste(img, (int(cfg["x"] - cfg["r"]), int(cfg["y"] - cfg["r"])), img)

    # Main picture
    if _CIRCLES_OPTIONS["add_watermark"] and _CIRCLES_OPTIONS["add_date"]:
        vertical_displace = 0.04
    else:
        vertical_displace = 0.0

    add_picture_to_image(
        main_profile_picture,
        dict(
            x=image_size / 2,
            y=(1 + vertical_displace) * image_size / 2,
            r=image_size * _CIRCLES_RADIUSES[n_orbits][0],
        ),
    )

    # Pictures
    for orbit_index, orbit in enumerate(_CIRCLES_ORBITS_DEF):
        orbit_pictures = orbit["pictures"]
        angle_step = 360 / orbit["count"]
        for i in range(orbit["count"]):
            if i >= len(orbit_pictures):
                break

            t = (i * angle_step + orbit_index * 30) * (math.pi / 180)  # in radians
            add_picture_to_image(
                orbit_pictures[i],
                dict(
                    x=math.cos(t) * image_size * orbit["distance"] + image_size / 2,
                    y=math.sin(t) * image_size * orbit["distance"] + (1 + vertical_displace) * image_size / 2,
                    r=image_size * orbit["radius"],
                ),
            )

    return cv.resize((image_size // 3, image_size // 3), Image.Resampling.LANCZOS)
